File: android-native/app/src/main/java/com/openflix/presentation/screens/livetv/MultiviewViewModel.kt
package com.openflix.presentation.screens.livetv

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openflix.domain.model.Channel
import com.openflix.data.repository.LiveTVRepository
import com.openflix.player.MultiviewPlayer
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * ViewModel for YouTube TV-style Multiview.
 *
 * Key behaviors:
 * - Audio follows focus: D-pad to a view = audio switches automatically
 * - Fullscreen toggle: OK press expands focused view, Back returns to grid
 * - Fixed 2x2 grid layout (no layout cycling)
 * - Channel picker on long-press only
 */
@HiltViewModel
class MultiviewViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val liveTVRepository: LiveTVRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(MultiviewUiState())
    val uiState: StateFlow<MultiviewUiState> = _uiState.asStateFlow()

    // Player instances for each slot
    private val players = mutableMapOf<Int, MultiviewPlayer>()

    init {
        loadChannels()
    }

    private fun loadChannels() {
        viewModelScope.launch {
            liveTVRepository.getChannels()
                .onSuccess { channels ->
                    _uiState.update { it.copy(allChannels = channels) }
                }
                .onFailure { e ->
                    Timber.e(e, "Failed to load channels for multiview")
                }
        }
    }

    fun initializeSlots(initialChannelIds: List<String>) {
        val channels = _uiState.value.allChannels
        if (channels.isEmpty()) {
            Timber.w("No channels available for multiview")
            return
        }

        val initialChannels = initialChannelIds.mapNotNull { id ->
            channels.find { it.id == id }
        }.take(4)

        // Default to 4 channels for YTTV-style 2x2 grid
        val channelsToUse = if (initialChannels.size >= 2) {
            initialChannels
        } else {
            channels.take(4)
        }

        val slots = channelsToUse.mapIndexed { index, channel ->
            val player = MultiviewPlayer(context).also { it.initialize() }
            players[index] = player
            channel.streamUrl?.let { url -> player.play(url) }

            MultiviewSlot(
                index = index,
                channel = channel,
                isReady = false
            )
        }

        // Set audio to first slot (audio follows focus, focus starts at 0)
        players[0]?.unmute()

        _uiState.update {
            it.copy(
                slots = slots,
                focusedSlotIndex = 0,
                audioSlotIndex = 0
            )
        }

        Timber.d("Multiview initialized with ${slots.size} slots (YTTV-style)")
    }

    fun getPlayer(slotIndex: Int): MultiviewPlayer? = players[slotIndex]

    /**
     * YTTV behavior: focus change = audio follows.
     * When D-pad moves to a slot, audio switches to it automatically.
     */
    fun setFocusedSlot(index: Int) {
        if (index == _uiState.value.focusedSlotIndex) return
        if (index >= _uiState.value.slots.size) return

        // Audio follows focus
        switchAudioToSlot(index)

        _uiState.update {
            it.copy(
                focusedSlotIndex = index,
                showOverlay = true
            )
        }
    }

    private fun switchAudioToSlot(slotIndex: Int) {
        players.forEach { (index, player) ->
            if (index == slotIndex) player.unmute() else player.mute()
        }
        _uiState.update { it.copy(audioSlotIndex = slotIndex) }
        Timber.d("Audio switched to slot $slotIndex")
    }

    /**
     * YTTV behavior: OK press = enter fullscreen for focused slot.
     */
    fun enterFullscreen() {
        val focused = _uiState.value.focusedSlotIndex
        _uiState.update {
            it.copy(
                isFullscreen = true,
                fullscreenSlotIndex = focused
            )
        }
        Timber.d("Entered fullscreen on slot $focused")
    }

    /**
     * YTTV behavior: Back from fullscreen = return to 2x2 grid.
     */
    fun exitFullscreen() {
        _uiState.update {
            it.copy(
                isFullscreen = false,
                fullscreenSlotIndex = null
            )
        }
        Timber.d("Exited fullscreen, back to grid")
    }

    fun changeChannelInSlot(slotIndex: Int, direction: Int) {
        val state = _uiState.value
        val slot = state.slots.getOrNull(slotIndex) ?: return
        val allChannels = state.allChannels
        if (allChannels.isEmpty()) return

        val currentIndex = allChannels.indexOfFirst { it.id == slot.channel.id }
        var newIndex = currentIndex + direction
        if (newIndex < 0) newIndex = allChannels.size - 1
        if (newIndex >= allChannels.size) newIndex = 0

        val newChannel = allChannels[newIndex]

        players[slotIndex]?.apply {
            newChannel.streamUrl?.let { url -> play(url) }
        }

        val updatedSlots = state.slots.toMutableList()
        updatedSlots[slotIndex] = slot.copy(channel = newChannel, isReady = false)

        _uiState.update { it.copy(slots = updatedSlots, showOverlay = true) }
        Timber.d("Changed slot $slotIndex to channel: ${newChannel.name}")
    }

    fun swapChannel(slotIndex: Int, newChannel: Channel) {
        val state = _uiState.value
        val slot = state.slots.getOrNull(slotIndex) ?: return

        players[slotIndex]?.apply {
            newChannel.streamUrl?.let { url -> play(url) }
        }

        val updatedSlots = state.slots.toMutableList()
        updatedSlots[slotIndex] = slot.copy(channel = newChannel, isReady = false)

        _uiState.update { it.copy(slots = updatedSlots) }
        Timber.d("Swapped slot $slotIndex to channel: ${newChannel.name}")
    }

    fun showOverlay() {
        _uiState.update { it.copy(showOverlay = true) }
    }

    fun hideOverlay() {
        _uiState.update { it.copy(showOverlay = false) }
    }

    fun showControls() {
        _uiState.update { it.copy(showControls = true) }
    }

    fun hideControls() {
        _uiState.update { it.copy(showControls = false) }
    }

    fun showChannelPicker(slotIndex: Int) {
        _uiState.update { it.copy(channelPickerSlotIndex = slotIndex) }
    }

    fun hideChannelPicker() {
        _uiState.update { it.copy(channelPickerSlotIndex = null) }
    }

    fun startOverSlot(slotIndex: Int) {
        val state = _uiState.value
        val slot = state.slots.getOrNull(slotIndex) ?: return

        viewModelScope.launch {
            liveTVRepository.getStartOverInfo(slot.channel.id)
                .onSuccess { startOverInfo ->
                    if (startOverInfo.available && !startOverInfo.streamUrl.isNullOrBlank()) {
                        players[slotIndex]?.apply {
                            play(startOverInfo.streamUrl)
                        }

                        val updatedSlots = state.slots.toMutableList()
                        updatedSlots[slotIndex] = slot.copy(
                            isTimeshifted = true,
                            timeshiftProgramTitle = startOverInfo.programTitle
                        )
                        _uiState.update { it.copy(slots = updatedSlots) }
                        Timber.d("Started over program '${startOverInfo.programTitle}' in slot $slotIndex")
                    }
                }
                .onFailure { e ->
                    Timber.e(e, "Failed to get start over info for slot $slotIndex")
                }
        }
    }

    fun goLiveSlot(slotIndex: Int) {
        val state = _uiState.value
        val slot = state.slots.getOrNull(slotIndex) ?: return

        players[slotIndex]?.apply {
            slot.channel.streamUrl?.let { url -> play(url) }
        }

        val updatedSlots = state.slots.toMutableList()
        updatedSlots[slotIndex] = slot.copy(
            isTimeshifted = false,
            timeshiftProgramTitle = null
        )
        _uiState.update { it.copy(slots = updatedSlots) }
        Timber.d("Went live in slot $slotIndex")
    }

    override fun onCleared() {
        super.onCleared()
        players.values.forEach { it.release() }
        players.clear()
        Timber.d("MultiviewViewModel cleared, all players released")
    }
}

data class MultiviewUiState(
    val slots: List<MultiviewSlot> = emptyList(),
    val allChannels: List<Channel> = emptyList(),
    val focusedSlotIndex: Int = 0,
    val audioSlotIndex: Int = 0,
    val isFullscreen: Boolean = false,
    val fullscreenSlotIndex: Int? = null,
    val showOverlay: Boolean = true,
    val showControls: Boolean = true,
    val channelPickerSlotIndex: Int? = null
)

data class MultiviewSlot(
    val index: Int,
    val channel: Channel,
    val isReady: Boolean = false,
    val isMuted: Boolean = true,
    val isTimeshifted: Boolean = false,
    val timeshiftProgramTitle: String? = null
)

enum class MultiviewLayout {
    SINGLE,
    TWO_BY_ONE,
    ONE_BY_TWO,
    THREE_GRID,
    TWO_BY_TWO
}

File: OpenFlix-tvOS/OpenFlix-tvOS/ViewModels/MultiviewViewModel.swift
import Foundation
import AVKit
import Combine

/// Layout options for multiview
enum MultiviewLayout: String, CaseIterable {
    case single = "single"
    case twoByOne = "2x1"      // Side by side
    case oneByTwo = "1x2"      // Stacked
    case threeGrid = "2+1"     // 2 top, 1 bottom
    case twoByTwo = "2x2"      // 4 quadrants

    var maxSlots: Int {
        switch self {
        case .single: return 1
        case .twoByOne, .oneByTwo: return 2
        case .threeGrid: return 3
        case .twoByTwo: return 4
        }
    }

    var icon: String {
        switch self {
        case .single: return "rectangle"
        case .twoByOne: return "rectangle.split.2x1"
        case .oneByTwo: return "rectangle.split.1x2"
        case .threeGrid: return "rectangle.split.2x2"
        case .twoByTwo: return "rectangle.split.2x2.fill"
        }
    }
}

/// Represents one slot in the multiview grid
struct MultiviewSlot: Identifiable {
    let id = UUID()
    var index: Int
    var channel: Channel
    var player: AVPlayer?
    var isReady: Bool = false
    var isMuted: Bool = true
    var isTimeshifted: Bool = false
    var timeshiftProgramTitle: String?
    var isBuffering: Bool = false
}

@MainActor
class MultiviewViewModel: ObservableObject {
    private let liveTVRepository = LiveTVRepository()

    @Published var slots: [MultiviewSlot] = []
    @Published var allChannels: [Channel] = []
    @Published var layout: MultiviewLayout = .twoByOne
    @Published var focusedSlotIndex: Int = 0
    @Published var showControls: Bool = true
    @Published var channelPickerSlotIndex: Int? = nil
    @Published var isLoading = false
    @Published var error: String?

    private var controlsHideTask: Task<Void, Never>?

    // MARK: - Initialization

    func loadChannels() async {
        isLoading = true
        defer { isLoading = false }

        do {
            try await liveTVRepository.loadChannels()
            allChannels = liveTVRepository.channels
        } catch {
            self.error = error.localizedDescription
        }
    }

    func initializeSlots(initialChannelIds: [String] = []) {
        guard !allChannels.isEmpty else { return }

        // Find initial channels or use first few
        var channelsToUse: [Channel] = []

        for id in initialChannelIds {
            if let channel = allChannels.first(where: { $0.id == id }) {
                channelsToUse.append(channel)
            }
        }

        // If no initial channels, use first 2
        if channelsToUse.isEmpty {
            channelsToUse = Array(allChannels.prefix(2))
        }

        // Create slots
        slots = channelsToUse.enumerated().map { index, channel in
            var slot = MultiviewSlot(index: index, channel: channel)
            slot.player = createPlayer(for: channel)
            return slot
        }

        // Set layout based on slot count
        layout = layoutForSlotCount(slots.count)

        // Start playing
        for i in slots.indices {
            startPlaying(slot: i)
        }
    }

    // MARK: - Player Management

    private func createPlayer(for channel: Channel) -> AVPlayer? {
        guard let streamUrl = channel.streamUrl, let url = URL(string: streamUrl) else {
            return nil
        }

        let player = AVPlayer(url: url)
        player.isMuted = true // Start muted
        return player
    }

    private func startPlaying(slot index: Int) {
        guard index < slots.count else { return }
        slots[index].player?.play()
        slots[index].isReady = true
    }

    func getPlayer(for slotIndex: Int) -> AVPlayer? {
        guard slotIndex < slots.count else { return nil }
        return slots[slotIndex].player
    }

    // MARK: - Focus Management

    func setFocusedSlot(_ index: Int) {
        guard index >= 0 && index < slots.count else { return }
        focusedSlotIndex = index
    }

    // MARK: - Channel Surfing

    func changeChannelInSlot(_ slotIndex: Int, direction: Int) {
        guard slotIndex < slots.count else { return }

        let currentChannel = slots[slotIndex].channel
        guard let currentIndex = allChannels.firstIndex(where: { $0.id == currentChannel.id }) else { return }

        var newIndex = currentIndex + direction
        if newIndex < 0 { newIndex = allChannels.count - 1 }
        if newIndex >= allChannels.count { newIndex = 0 }

        let newChannel = allChannels[newIndex]
        swapChannel(slotIndex, newChannel: newChannel)
    }

    func swapChannel(_ slotIndex: Int, newChannel: Channel) {
        guard slotIndex < slots.count else { return }

        // Stop old player
        slots[slotIndex].player?.pause()

        // Create new player
        slots[slotIndex].channel = newChannel
        slots[slotIndex].player = createPlayer(for: newChannel)
        slots[slotIndex].isReady = false
        slots[slotIndex].isTimeshifted = false
        slots[slotIndex].timeshiftProgramTitle = nil

        // Start playing
        startPlaying(slot: slotIndex)
    }

    // MARK: - Slot Management

    func addSlot() {
        guard slots.count < 4 else { return }

        // Find a channel not already in use
        let usedChannelIds = Set(slots.map { $0.channel.id })
        guard let newChannel = allChannels.first(where: { !usedChannelIds.contains($0.id) }) ?? allChannels.first else {
            return
        }

        var slot = MultiviewSlot(index: slots.count, channel: newChannel)
        slot.player = createPlayer(for: newChannel)
        slots.append(slot)

        // Update layout
        layout = layoutForSlotCount(slots.count)

        // Start playing
        startPlaying(slot: slots.count - 1)
    }

    func removeSlot(_ slotIndex: Int) {
        guard slots.count > 1, slotIndex < slots.count else { return }

        // Stop and release player
        slots[slotIndex].player?.pause()
        slots[slotIndex].player = nil

        // Remove slot
        slots.remove(at: slotIndex)

        // Reindex remaining slots
        for i in slots.indices {
            slots[i].index = i
        }

        // Update layout
        layout = layoutForSlotCount(slots.count)

        // Adjust focus if needed
        if focusedSlotIndex >= slots.count {
            focusedSlotIndex = slots.count - 1
        }
    }

    // MARK: - Audio Control

    func toggleMuteOnSlot(_ slotIndex: Int) {
        guard slotIndex < slots.count else { return }

        let currentlyMuted = slots[slotIndex].isMuted

        // If unmuting this slot, mute all others first
        if currentlyMuted {
            for i in slots.indices {
                slots[i].isMuted = true
                slots[i].player?.isMuted = true
            }
        }

        // Toggle this slot
        slots[slotIndex].isMuted = !currentlyMuted
        slots[slotIndex].player?.isMuted = !currentlyMuted
    }

    func setAudioFocus(_ slotIndex: Int) {
        guard slotIndex < slots.count else { return }

        // Mute all slots
        for i in slots.indices {
            slots[i].isMuted = true
            slots[i].player?.isMuted = true
        }

        // Unmute focused slot
        slots[slotIndex].isMuted = false
        slots[slotIndex].player?.isMuted = false
    }

    // MARK: - Layout Control

    func cycleLayout() {
        let layouts = MultiviewLayout.allCases.filter { $0.maxSlots >= slots.count }
        guard let currentIndex = layouts.firstIndex(of: layout) else {
            layout = layouts.first ?? .twoByOne
            return
        }

        let nextIndex = (currentIndex + 1) % layouts.count
        layout = layouts[nextIndex]
    }

    private func layoutForSlotCount(_ count: Int) -> MultiviewLayout {
        switch count {
        case 1: return .single
        case 2: return .twoByOne
        case 3: return .threeGrid
        default: return .twoByTwo
        }
    }

    // MARK: - Timeshift / Start Over

    func startOverSlot(_ slotIndex: Int) async {
        guard slotIndex < slots.count else { return }

        let channel = slots[slotIndex].channel

        do {
            let startOverInfo = try await liveTVRepository.getStartOverInfo(channelId: channel.id)

            if startOverInfo.available, let streamUrlString = startOverInfo.streamUrl,
               let url = URL(string: streamUrlString) {
                // Stop current player
                slots[slotIndex].player?.pause()

                // Create new player with timeshift URL
                slots[slotIndex].player = AVPlayer(url: url)
                slots[slotIndex].isTimeshifted = true
                slots[slotIndex].timeshiftProgramTitle = startOverInfo.programTitle

                // Start playing
                slots[slotIndex].player?.play()
                slots[slotIndex].player?.isMuted = slots[slotIndex].isMuted
            }
        } catch {
            // Silently fail
        }
    }

    func goLiveSlot(_ slotIndex: Int) {
        guard slotIndex < slots.count else { return }

        let channel = slots[slotIndex].channel

        // Stop current player
        slots[slotIndex].player?.pause()

        // Create new player with live URL
        slots[slotIndex].player = createPlayer(for: channel)
        slots[slotIndex].isTimeshifted = false
        slots[slotIndex].timeshiftProgramTitle = nil

        // Start playing
        startPlaying(slot: slotIndex)
    }

    // MARK: - Controls Visibility

    func showControlsTemporarily() {
        showControls = true
        scheduleHideControls()
    }

    func hideControlsNow() {
        showControls = false
    }

    private func scheduleHideControls() {
        controlsHideTask?.cancel()
        controlsHideTask = Task {
            try? await Task.sleep(nanoseconds: 8_000_000_000) // 8 seconds
            if !Task.isCancelled {
                showControls = false
            }
        }
    }

    // MARK: - Channel Picker

    func showChannelPicker(for slotIndex: Int) {
        channelPickerSlotIndex = slotIndex
    }

    func hideChannelPicker() {
        channelPickerSlotIndex = nil
    }

    // MARK: - Cleanup

    func cleanup() {
        controlsHideTask?.cancel()

        for slot in slots {
            slot.player?.pause()
        }
        slots.removeAll()
    }
}
